common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -242,15 +248,54 @@ internal class WebSocketCoreImpl(
                             }
                         }
                     }
+                    if (!didConnect) {
+                        urlObserverJob?.cancel()
+                    }
                     didConnect
                 } catch (e: Exception) {
                     Timber.e(e, "Unable to authenticate")
+                    urlObserverJob?.cancel()
                     false
                 }
             }
         }
     }
 
+    /**
+     * Starts observing URL changes using a single flow subscription.
+     * Returns the first available URL and continues observing for changes in the background.
+     * When the URL changes, disconnects immediately the WebSocket to trigger reconnection via [handleClosingSocket].
+     *
+     * @return the first URL if available, or `null` if not
+     */
+    private suspend fun startUrlObserverAndAwaitFirstUrl(): URL? {
+        urlObserverJob?.cancel()
+
+        val firstUrlDeferred = CompletableDeferred<URL?>()
+
+        urlObserverJob = wsScope.launch {
+            var isFirstEmission = true
+            connectionStateProvider().urlFlow()
+                .collect { urlState ->
+                    val url = (urlState as? UrlState.HasUrl)?.url
+                    if (isFirstEmission) {
+                        isFirstEmission = false
+                        firstUrlDeferred.complete(url)
+                    } else if (url != connectedUrl) {
+                        if (urlState is UrlState.InsecureState) {
+                            Timber.w("Insecure state, disconnecting immediately.")
+                        } else {
+                            Timber.w("URL changed, disconnecting immediately.")
+                        }
+                        // we abruptly cancel the connection since we might be in an insecure state
+                        connection?.cancel()
```

Comment:
- Maybe introduce `WebSocketState.CLOSED_URL_CHANGE` for this?
- In case the URL changes, do we want the same delay for reconnecting (10s) or none / a shorter delay?